### PR TITLE
Add random sleep for zero copy

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add random sleep before merges/mutations execution to split load more evenly between replicas in case of zero-copy replication.

### CODE REVIEW

File: src/Storages/MergeTree/MergeFromLogEntryTask.cpp
Comment (serxa) [inline src/Storages/MergeTree/MergeFromLogEntryTask.cpp:227]:
```suggestion
                /// Here we are trying to mitigate the skew of merges execution because of faster/slower replicas.
```

File: src/Storages/MergeTree/MergeFromLogEntryTask.cpp
Comment (serxa) [inline src/Storages/MergeTree/MergeFromLogEntryTask.cpp:230]:
```suggestion
                /// the fast replica is not overloaded because amount of executing merges doesn't affect the ability to acquire locks for new merges.
```

File: src/Storages/MergeTree/MutateFromLogEntryTask.cpp
Comment (alesapin) [inline src/Storages/MergeTree/MutateFromLogEntryTask.cpp:145]:
```suggestion
                auto time_to_sleep_milliseconds = std::min(10000UL, std::uniform_int_distribution<uint64_t>(1, 1 + right_border_to_sleep_ms)(rng));
```

**Review by serxa — APPROVED** [review]
LGTM

File: src/Storages/MergeTree/MergeFromLogEntryTask.cpp
Comment (Algunenano) [inline src/Storages/MergeTree/MergeFromLogEntryTask.cpp:233]: I've been having a look at the code because we've been testing zero copy replication and seen too many requests to ZK during merge operations. The replicas that aren't doing the merge keep retrying the tasks as fast as possible (10x per second more or less) which is quite a bit.

This PR seems to help to address this (we are using 23.5, which doesn't have this change yet so I haven't tested it in a real scenario) but I have some concerns:
* The comment says that it `sleep random time up to 500ms for 1GB part and up to 7 seconds for 300GB part`. This doesn't appear to be true, but I feel it's a bug in the code. It's doing std::min with 10 seconds, so it's likely always sleeping for 10 seconds. Maybe it was meant to be a max?

* This is sleeping inside the task, which means that if you happen to have a bunch of tasks sleep here (for different tables or different merges) you will exhaust the background pools while doing nothing. IMO Ideally it should be doing some backoff outside the thread. We can't simply wait until merge is notified to ZK because that might never happen (the replica is restarted), but backoff seems better than constant checks.


What do you think, @alesapin?

Comment (serxa) [reply src/Storages/MergeTree/MergeFromLogEntryTask.cpp:233]: 10 seconds is the upper bound for really huge merges. `min` was intended. Also note that sleep is randomized from 1 ms to a value `k*log(merge_size)` So given this fact thread pool exhaustion is possible but less likely.

Comment (danipozo) [reply src/Storages/MergeTree/MergeFromLogEntryTask.cpp:233]: > Ideally it should be doing some backoff outside the thread.

It seems like the existence of the lock is [checked outside the thread](https://github.com/ClickHouse/ClickHouse/blob/4bb95305a477142baee094ec881e9dd0cbf5c2cc/src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp#L1379) too. The merging task is sent to the back of the queue without backoff though, AFAICT.

Comment (Algunenano) [reply src/Storages/MergeTree/MergeFromLogEntryTask.cpp:233]: Yep, your are right. I was misreading it and thinking about it the wrong way.

500 ms, which is the min, might still be too low in some cases though.

For example, checking the logs from a merge in 23.5 I have:
* Part size: 1.6GB
* Start time: 10:12:50.73
* End time: 10:14:54.28
* Task runs: 1252 => 10,09 runs/second

Estimated sleep time now: log(1.6GB/1GB) + 0.5 = 0.7 seconds
Estimated task runs: 124*((124/1252)+0.7) ~= 99 task runs

It is 10x better than before, but it seems still too many checks; specially since adding N replicas means also N checks. Waiting outside the thread and an exponential backoff still makes sense to me (for the future).

Comment (Algunenano) [reply src/Storages/MergeTree/MergeFromLogEntryTask.cpp:233]: Fixed a bunch of the numbers. The estimation is that now there would be 276 runs instead of 1252 runs. So around 2 per second in this case. Way too many IMO.

Comment (Algunenano) [reply src/Storages/MergeTree/MergeFromLogEntryTask.cpp:233]: Related to this, the problem were tasks were retried too fast happens also in part because in our tests we use a  multidisk policy (local hot storage and cloud cold storage) and in that case CH doesn't wait until the zero copy lock is available to retry.

Having the sleeps helps in that case but it's not the main reason why they were introduced (split merge load), so I've proposed a change in behaviour in that case in https://github.com/ClickHouse/ClickHouse/pull/52376

**Pull request merged**
